### PR TITLE
Allow separate launch composition

### PR DESCRIPTION
**Pull request by eric1221bday**
As mentioned in #75, this is my attempt to allow composing nodes in different launch files. This change allows this by making target_container also a substitution so users can pass in the container node name. That way the container node itself can be executed first in another launch file.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ __pycache__
 .mypy_cache
 .pytest_cache
 .flake8
+.vscode
```
[the remark this answers is not in the document]
Comment (mjcarroll) [reply .gitignore]: It would seem safer to add it to your global .gitignore if VSCode is your editor of choice.  I believe @dirk-thomas uses VSCode, is that the approach that you use?

[the remark this answers is not in the document]
Comment (dirk-thomas) [reply .gitignore]: Since there is an almost infinite amount of possible editors and custom config files, yes, I think those should be ignored by the user globally.

File: launch_ros/launch_ros/actions/load_composable_nodes.py [hunk 2]
```diff
@@ -55,15 +58,10 @@ def __init__(
         :param composable_node_descriptions: descriptions of composable nodes to be loaded
         :param target_container: the container to load the nodes into
         """
-        ensure_argument_type(
-            target_container,
-            ComposableNodeContainer,
-            'target_container',
-            'LoadComposableNodes'
-        )
```
Comment (hidmic) [inline launch_ros/launch_ros/actions/load_composable_nodes.py]: @eric1221bday why dropping the type check? `ensure_argument_type` can take more than one type.

File: launch_ros/launch_ros/actions/load_composable_nodes.py [hunk 3]
```diff
@@ -39,7 +42,7 @@ def __init__(
         self,
         *,
         composable_node_descriptions: List[ComposableNode],
-        target_container: ComposableNodeContainer,
+        target_container: Union[SomeSubstitutionsType, ComposableNodeContainer],
```
[the remark this answers is not in the document]
Comment (hidmic) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: Actually, I'd rather have a single argument. In the most general case a user would have to provide a fully qualified node name (FQNN), but in Python we offer the chance to pass in a `ComposableNodeContainer` action from which an FQNN can be fetched. The later and current behavior is a bit of a heuristic, as the underlying node may ignore all name remappings.

File: launch_ros/launch_ros/actions/load_composable_nodes.py [hunk 4]
```diff
@@ -161,10 +159,17 @@ def execute(
         context: LaunchContext
     ) -> Optional[List[Action]]:
         """Execute the action."""
+        # resolve target container node name
+        if isinstance(self.__target_container, ComposableNodeContainer):
+            self.__final_target_container_name = self.__target_container.node_name
+        else:  # TODO: maybe should do better type checking here
+            subs = normalize_to_list_of_substitutions(self.__target_container)
+            self.__final_target_container_name = perform_substitutions(context, subs)
```
Comment (hidmic) [inline launch_ros/launch_ros/actions/load_composable_nodes.py]: @eric1221bday why not resolving this TODO now?

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 2 above)
Comment (eric1221bday) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: @hidmic It's because it has trouble with Union types, which is what SubstitutionType ultimately is. Alternative we can list out every single type a substitution can possibly be, but that felt very hacky to me.

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eric1221bday) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: My personal thought when designing this was to reduce impact on current API, hence why a union type. I also started off thinking it should've been a separate argument. I agree that this makes the argument name vague with dual meaning, but I think this can be handled via better docstrings and docs in general.

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 4 above)
Comment (eric1221bday) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: @hidmic So I meant to revisit this, the core of the issue is the same reason I removed the ensure_argument_type at the top, which is that Unions are tricky to type check. There's a stackoverflow [here](https://stackoverflow.com/a/49471187) with a suggested way of checking using both the `__origin__` and `__args__` attribute, but I'm unsure how advisable that is.

File: .gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (eric1221bday) [reply .gitignore]: removed

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wjwwood) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: I like one argument that can either be a string (substitution) or a container node class. That way the two ways of specifying are mutually exclusive by design.

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 2 above)
Comment (wjwwood) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: You can use `SomeSubstitutionsType_types_tuple`, e.g.:

https://github.com/ros2/launch/blob/8f1ed681973aad9b48ecc6712e54d2f66ee4dac3/launch/launch/actions/timer_action.py#L66

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ivanpauno) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: ok, so one argument then.

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 4 above)
Comment (wjwwood) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: As I said above you can use the types tuple we have, but if you choose to leave the TODO, please put your github handle into it (that's our convention): `# TODO(eric1221bday): ...`

File: launch_ros/launch_ros/actions/load_composable_nodes.py [hunk 5]
```diff
@@ -161,10 +159,17 @@ def execute(
         context: LaunchContext
     ) -> Optional[List[Action]]:
         """Execute the action."""
+        # resolve target container node name
+        if isinstance(self.__target_container, ComposableNodeContainer):
```
Comment (wjwwood) [inline launch_ros/launch_ros/actions/load_composable_nodes.py]: Will this work if someone were to subclass `ComposableNodeContainer`? Ideally it would. I don't remember what the gotcha's are there, but I know we have this utility for some reason: https://github.com/ros2/launch/blob/a46aabdde1f58c7e2e5c056c0c4a1ea830faac5c/launch/launch/utilities/class_tools_impl.py#L34 which uses `issubclass()` internally.

Comment (eric1221bday) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: @wjwwood so at least according to [this](https://stackoverflow.com/a/1549854) it seems like it should, but I'm happy to use the created util function instead

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 2 above)
Comment (eric1221bday) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: @wjwwood added it back in with suggested tuple, seems to work

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 4 above)
Comment (eric1221bday) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: @wjwwood alright I ended up using `is_a_subclass` for the ComposableNodeContainer and `isinstance` with `SomeSubstitutionsType_types_tuple` to do the type checking. If the argument is neither I will log and error and return. Considering the type checking at the constructor this is a bit of an unlikely hit, but no harm really.

File: launch_ros/launch_ros/actions/load_composable_nodes.py (hunk 5 above)
Comment (eric1221bday) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: As mentioned above, ended but using `is_a_subclass`

Comment (ivanpauno) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: @wjwwood I don't understand `is_a_subclass` implementation.
[isinstance](https://docs.python.org/2/library/functions.html#isinstance) already returns `True` if the object is instance of a subclass.
[issubclass](https://docs.python.org/2/library/functions.html#issubclass) returns `True` if a class is a subclass of another (and not when an object is instance of a subclass).
So I don't understand why `is_a_subclass` checks for both `isinstance` and `issubclass`.

IMO, using [is_a](https://github.com/ros2/launch/blob/a46aabdde1f58c7e2e5c056c0c4a1ea830faac5c/launch/launch/utilities/class_tools_impl.py#L25) util in launch or `isinstance` is ok.

Comment (wjwwood) [reply launch_ros/launch_ros/actions/load_composable_nodes.py]: @ivanpauno Yeah, I don't remember a rationale off-hand, so maybe your recommendation is sufficient.

**Comment by eric1221bday** [comment]
@ivanpauno so is this good?

**Comment by ivanpauno** [comment]
> @ivanpauno so is this good?

Yes, it's good.
We can follow-up about `is_a_subclass` usage later (https://github.com/ros2/launch_ros/pull/77#discussion_r328781278).

----

We're currently in a feature freeze, so this is not going to be merged into `master` until we create an Eloquent branch (probably in one/two weeks).

**Comment by eric1221bday** [comment]
@ivanpauno gentle ping now that the feature freeze seems to be over.

File: .gitignore [hunk 6]
```diff
@@ -4,3 +4,4 @@ __pycache__
 .mypy_cache
 .pytest_cache
 .flake8
+
```
Comment (ivanpauno) [inline .gitignore]: nit: delete unnecessary change.

**Comment by ivanpauno** [comment]
@eric1221bday please address [flake8 failures](https://ci.ros2.org/job/ci_linux/8764/testReport/junit/launch_ros.test/test_flake8/test_flake8/).

**Comment by eric1221bday** [comment]
@ivanpauno should be fixed

**Comment by ivanpauno** [comment]
Sanity check (using rebased branch `ivanpauno/test-#77`):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8793/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4613/)
* MacOS [Build Status](http://ci.ros2.org/job/ci_osx/7207/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8713/)

**Comment by ivanpauno** [comment]
Going in! Thanks for the contribution @eric1221bday !

**Comment by jacobperron** [comment]
I'd like to see this change backported to Eloquent.

Since it looks like mainly bug fixes since the last release, maybe we can sync the eloquent branch with master?

**Comment by ivanpauno** [comment]
> Since it looks like mainly bug fixes since the last release, maybe we can sync the eloquent branch with master?

Sounds reasonable to me.
